### PR TITLE
fix debug_storageRangeAt

### PR DESCRIPTION
**Pull request by bgelb**

### CODE REVIEW

File: turbo/transactions/tracing.go [hunk 1]
```diff
@@ -65,7 +65,12 @@ func ComputeTxEnv(ctx context.Context, block *types.Block, cfg *params.ChainConf
 		}
 		// Ensure any modifications are committed to the state
 		// Only delete empty objects if EIP158/161 (a.k.a Spurious Dragon) is in effect
-		_ = statedb.FinalizeTx(vmenv.ChainRules(), state.NewNoopWriter())
+		_ = statedb.FinalizeTx(vmenv.ChainRules, reader)
+
+		if idx+1 == len(block.Transactions()) {
+			// Return the state from evaluating all txs in the block, note no msg or TxContext in this case
+			return nil, BlockContext, vm.TxContext{}, statedb, reader, nil
+		}
```
Comment (AskAlexSharov) [inline turbo/transactions/tracing.go:73]: This line doing same? https://github.com/ledgerwatch/erigon/pull/3148/files#diff-664a6f9734f5a173f4805bd1eee6763dab47fff1aec9839fdbef67036d7b09faR58

Comment (bgelb) [reply turbo/transactions/tracing.go:73]: sorry I did not quite understand this comment.

**Comment by AlexeyAkhunov** [comment]
Thank you for this PR! Does it solve your issue?

**Comment by bgelb** [comment]
> Thank you for this PR! Does it solve your issue?

Yes. We are using in our application now and seems to do what we need.

File: turbo/transactions/tracing.go (hunk 1 above)
Comment (AskAlexSharov) [reply turbo/transactions/tracing.go:73]: nevermind

**Comment by AskAlexSharov** [comment]
a bit weird that our PlainState class, does remember storage changes, but loosing all other changes...
IntraBlocksState must remember all changes (including storage), for me it's unclear why replacing NoopWriter by PlainState fixed something (smells like "fixed by accident because of even amount of bugs").

**Comment by bgelb** [comment]
> a bit weird that our PlainState class, does remember storage changes, but loosing all other changes... IntraBlocksState must remember all changes (including storage), for me it's unclear why replacing NoopWriter by PlainState fixed something (smells like "fixed by accident because of even amount of bugs").

The reason is that the storage within PlainState is what is used to generate the result of the storageRangeAt call, see:
https://github.com/ledgerwatch/erigon/blob/c913f35c2e42e9cc1a1c31d681f834531cc095bb/cmd/rpcdaemon/commands/storage_range.go#L30

If that storage is not updated, then it reflects the storage as of the beginning of the block, no matter what transaction index is specified, because the storage updates are never committed into the PlainState object ("stateReader").

**Comment by AskAlexSharov** [comment]
now I see. then it make more sense. 
Please merge devel to your branch and we can merge this PR.

**Comment by bgelb** [comment]
> now I see. then it make more sense. Please merge devel to your branch and we can merge this PR.

I have merged latest devel.

**Comment by bgelb** [comment]
> > now I see. then it make more sense. Please merge devel to your branch and we can merge this PR.
> 
> I have merged latest devel.

It seems that somehow the change to link

> now I see. then it make more sense. Please merge devel to your branch and we can merge this PR.

It appears that the change from NoopWriter -> PlainState (on line 68) did not get merged properly. This is a critical element of the fix. What is the right way to proceed to fix this issue?

**Pull request merged**